Update mc readers

andLaing opened the pull request:
Updates the pandas DataFrame based readers
for the MC hits, particles and sensors taking into
account the new nexus output structure.

Adapts existing functions to return the same structure
for both old and new format files.

andLaing committed:
Change fixture KrMC_true_hits to use DF based readers

This fixture used load_mchits, changed to use
load_mchits_df and related tests adjusted to
expect this structure

andLaing committed:
Patch test_voxels_with_no_hits for load_mchits_df

mmkekic commented on invisible_cities/reco/paolina_functions_test.py:
```diff
@@ -239,20 +239,15 @@ def test_voxels_with_no_hits(ICDATADIR):
+    hits_df  = load_mchits_df(hit_file)
+    ## Patch to BHits, temporary?
+    evt_hits = hits_df.loc[evt_number][['x', 'y', 'z', 'energy']].values
+    bhit_seq = [BHit(*hit) for hit in evt_hits]
```
I think we should have an official function that casts mchits_df to a dictionary (the output of load_mchits) should be in mcinfo_io

mmkekic commented on invisible_cities/cities/penthesilea_test.py:
```diff
@@ -195,15 +197,15 @@ def test_penthesilea_produces_mcinfo(KrMC_pmaps_filename, KrMC_hdst, config_tmpd
-
-        assert len(penthesilea_hits) == len(true_hits)
-        (assert_MChit_equality(p_hit, t_hit) for p_hit, t_hit in zip(penthesilea_hits, true_hits))
+    assert np.all(penthesilea_evts.index.levels[0] == true_evts.index.levels[0])
```
Since you are reading all as dataframes now isnt `assert_dataframes_equal(penthesilea_evts, KrMC_true_hits.hdst)` enough?
Also maybe use the `assert_dataframes_equal` from our testing utils?

andLaing replied:
Is there any reason to use our own assert_dataframes_equal and not `pandas.testing.assert_frame_equal`?

mmkekic replied:
Maybe to be consistent with (mostly?) all other tests?  But also it might be easier to track in case something changes in future pandas/numpy version (ie, the testing used to be in pandas.utils.testing and now is pandas.testing...)
Also shouldn't assert close be more appropriate since those are floats?

andLaing committed:
improve test_penthesilea_true_hits_are_correct

Now uses pandas.testing.assert_frame_equal
to check all parts of frame without the
need to look over events

andLaing committed:
Change setup for test_voxels_with_no_hits to use cast function

andLaing committed:
Add database relevant arguments for copy_mc_info to remaining cities

arguments added to diomira

diomira cosmetics to improve readability

arguments added to irene

arguments added to penthesilea

arguments added to isidora

arguments added to esmeralda

andLaing committed:
Adapt diomira_test for new MC info copy

Checks event numbers and uses
pandas.testing.assert_frames_equal

Adapted test_diomira_read_multiple_files for new writer

Demonstrate test_diomira_exact_result passes without MC tables

diomira_test cosmetics

andLaing committed:
Adapt irene_test to use new mc readers

comparison of in and out now uses
pd.testing.assert_frame_equal

adapted test_irene_read_multiple_files for new MC

show test_irene_exact_result only fails due to MC changes

show test_irene_filters_empty_pmaps only fails due to MC changes

andLaing committed:
Adapt penthesilea_test for new MC

extents table not written in new mc output

Adapted test_penthesilea_read_multiple_files for new MC

Show test_penthesilea_exact_result only fails dur to MC change

Show test_penthesilea_exact_result only fails due to MC

andLaing committed:
Show test_isidora_exact_result only fails because of MC output

test_isidora_exact_result runs and writes MC
test_isidora_electrons_40keV fails due to the
test file electrons_40keV_z250_RWF.h5 having
an error in the extents table.

andLaing committed:
Adapt esmeralda_test for new MC

extents not written in new MC data writers

Adapted test_esmeralda_filters_events to use new MC readers

checks events using hits table rather than
extents table which is now not written

Show test_esmeralda_exact_result_all_events only failed due to MC

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -171,13 +175,73 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
 
-def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : List[int]=None):
-    """Copy from the input file to the output file the MC info of certain events.
+def save_mc_table(file_in : str                         ,
```
This seems rather complex, aren't we supposed to save all the output in a unique (new) MC format? ie, in copy_mc_info we read all relevant data (using our universal for both old and new readers) and just write the relevant tables? 
I would prefer to have output files with a unique structure even if some of the data was not previously saved in old MC format and we have to fill it with NaNs

andLaing replied:
It does output a unique format that's the same for all but you have to take into account that the input isn't always the same and output it in a readable format

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -186,26 +250,55 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+
+        with tb.open_file(file_name, 'r') as h5in:
+            mc_tbls  = ['hits', 'particles', 'sns_response']
+            evt_list = list(_try_unique_evt_itr(h5in.root.MC, mc_tbls))
```
do we understand why are there events that are not in particles or hits table?

andLaing replied:
This is a question for the NEXUS authors really.
There are certain types of simulations that have some and not other MC types saved, a simulation of optical photons for example would have an empty hits table as far as I understand.

paolafer replied:
There are different kinds of simulations, where we save different kinds of output (some will save only sensor response information and not true particles, sometimes the opposite happen, for instance). We want to be able to read all kind of MC files.

paolafer replied:
An option would be to resurrect the old `events` table, which contained two columns, namely the eventID and the total deposited true energy, but I don't know if it's worth it...

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -330,32 +456,68 @@ def read_mcparticles_df(h5in    : tb.file.File,
+
+def sensor_binning_old(file_name : str) -> pd.DataFrame:
+    # Is a pre-Feb 2020 MC file
+    config = pd.read_hdf(file_name, 'MC/configuration').set_index('param_key')
```
I think is better to use `dst_io.load_dst` ( that opens file using pytables) since we already saw that in different pandas version this might not work, depending on how file was created (ie pytables or pandas writer)

andLaing replied:
Do you have an example of a fail? I've never seen it fail in this context due to pandas versions.

mmkekic replied:
The beersheba_tests fail due to pandas not being able to read PSF file in version 0.25 (we use 0.23). This is an issue that needs to be solved but obviously not related to this PR

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -330,32 +456,68 @@ def read_mcparticles_df(h5in    : tb.file.File,
+def sensor_binning_old(file_name : str) -> pd.DataFrame:
+    # Is a pre-Feb 2020 MC file
+    config = pd.read_hdf(file_name, 'MC/configuration').set_index('param_key')
+    bins         = config[config.index.str.contains('binning')].copy()
```
since config was read in a previous line and not given as a mutable input, you don't have to copy it

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -330,32 +456,68 @@ def read_mcparticles_df(h5in    : tb.file.File,
+    return bins
+
+
+def get_sensor_types(file_name : str) -> pd.DataFrame:
```
I am confused between get, load and read names, can't we make a more systematic naming?

andLaing committed:
Adapt beersheba_test to use the modern copy_mc_info code

Removed remaining calls to old mc writers

remove assert MC/extents table in test_beersheba_contains_all_tables

Not copied in the new MC structure

Show test_beersheba_exact_result_joint failed due to new MC only

Show test_beersheba_exact_result_separate only failed due to MC

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -172,13 +181,97 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+    if 'MC' in h5out.root:
+        if 'event_mapping' in h5out.root.MC:
+            return h5out.root.MC.event_mapping.cols.file_index[-1]
+    return 0
```
I think here you should return -1 in case there is no table?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -172,13 +181,97 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+                   run_no  : Optional[int] = None) -> Dict:
+    """
+    Reads all the MC tables present in
+    file_in and sorts the Dataframes for
```
stores rather than sorts? also add arguments and return explanation in docstrings

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -172,13 +181,97 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+
+
+def read_mc_tables(file_in : str                 ,
+                   evt_arr : np.ndarray          ,
```
the evt_arr should be optional, if not given the whole table should be read

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -172,13 +181,97 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+    """
+    mctbls = get_mc_tbl_list(file_in)
+    if (len(evt_arr) > 0 and
+        len(np.intersect1d(evt_arr, get_event_numbers_in_file(file_in))) == 0):
```
I dont understand this check, why do you want to raise the error only if none of the required events are contained in the file but allow empty selection list? I think more appropriate  would be to raise either Error or Warning when some of the events are not present.

andLaing replied:
This was to reproduce the behaviour of that was present from PR #702 and copy the tables without event number as a variable in case of 0 events passing the cuts in the city being launched

mmkekic replied:
Remove this line, it was needed in PR #702 but note that previous writer worked on per-event basis. In this writer it really makes no sense to keep any check of this style

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -275,88 +443,134 @@ def load_mcparticles_df(file_name: str) -> pd.DataFrame:
+                              'initial_momentum_y': p_tb.col('momentum')[:, 1],
+                              'initial_momentum_z': p_tb.col('momentum')[:, 2],
+                              'kin_energy'        : p_tb.col('kin_energy'),
+                              'creator_proc'      : p_tb.col('creator_proc').astype('U20')})
```
I realized that in new format there are new columns:
`'length', 'final_momentum_z', 'final_momentum_x', 'final_momentum_y', 'final_proc'` 
(please @paolafer , @jmunozv , @jmalbos  confirm if this is new standard)
which means that we would never be able to mix the new and old format, ie the writer will not work, are we happy with that?

paolafer replied:
Yes, confirmed. I think that we are ok with not mixing new and old format.

mmkekic replied:
If we want to make interface blind to the file format maybe would be a better idea to fill the missing columns with nans? Cause someone doing analysis with old and new files would still have to have 2 versions of the code to work with each type, and we wanted to avoid that?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -275,88 +443,134 @@ def load_mcparticles_df(file_name: str) -> pd.DataFrame:
+                            left_index  =   True,
+                            right_index =   True,
+                            how         = 'left')
+        parts.rename(columns={"evt_number": "event_id"}, inplace = True)
```
no spaces around '=' in oneliners

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -187,26 +280,93 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+
+    evt_list = list(try_unique_evt_itr(h5in.root.MC, mc_tbls))
+    if len(evt_list) == 0:
+        raise tb.exceptions.NoSuchNodeError
```
this is a bit confusing, I am not sure that this function should raise an error if there are no tables with events or simply return an empty list?
In any case it is more clear to first get the name of the present tables and iterate over them extending the unique numpy array with `setdiff1d`or so

andLaing replied:
The exception was to maintain the expected try except structure approved for PR #702 but if you prefer I can look for ways to improve it?

We don't have access to the file name here to use the table list function and I'd prefer not to look at all tables in any case since many don't have an event_id column.

As to the extending, it seems like it'd need more gymnastics to do it that way since we'd have to initialise, right?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -172,13 +181,97 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+            tbl_dict[tbl] = evt_map[evt_map.event_id.isin(evt_arr)]
+        else                                         :
+            raise TypeError("MC table has no reader")
+    return tbl_dict
```
maybe we should also return a list of events?

andLaing replied:
To what end?

We send an event list in most user cases and if not it will give you the information for all events and the complete event list can be got by calling the get_event_numbers_in_file function

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -172,13 +181,97 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+                 db_file      : Optional[str]       = None,
+                 run_number   : Optional[int]       = None) -> None:
+    """
+    Copy from the input file to the output file the MC info of certain events.
```
description of `db_file` and `run_number`

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -187,26 +280,93 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+    """
+    with tb.open_file(file_name, 'r') as h5in:
+        mc_group = h5in.root.MC
+        return [MCTableType[tbl.name] for tbl in mc_group]
```
what happens if there is new table name in the file? Guess enum will complain, are we happy with the error?

andLaing replied:
I think so. All possible tables are there at the moment and have the same name as they do in file so if there's a change the Enum will give a readable error and the user will know that the code needs to be updated to allow for that table.

mmkekic replied:
In case there are tables in 'MC' group that are outside the enum class the `KeyError`will be raised. Maybe it needs a bit more explanation?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -187,26 +280,93 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+    config           = load_dst(file_name, 'MC', 'configuration')
+    if is_oldformat_file(file_name):
+        binning_keys = config.param_key.str.contains('binning')
+        new_names    = config[binning_keys].param_key.str.split(
```
this line is a bit hard to read, maybe a comment explains why and what we do?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -187,26 +280,93 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+                            run_no    : Optional[int] = None) -> pd.DataFrame:
+    if is_oldformat_file(file_name):
+        sns_pos = load_dst(file_name, 'MC', 'sensor_positions')
+        if sns_pos.shape[0] > 0:
```
if I understand this, the 'sensor_position' is *always* present in the old format file, but it might be empty? And if it is present the `db_file`and `run_no` are not optional and the code will crash. Do we need a protection against that?

andLaing replied:
I think the exception should be raised if the database info isn't sent with an old format file. Maybe I should change the ordering so it's raised even if the sensor_positions table is empty?

mmkekic replied:
The error is :
`TypeError: getattr(): attribute name must be string`
so I believe a more readable exception is in place.
Also I wonder if maybe we should throw a warning and return an empty DataFrame?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -187,26 +280,93 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+        if sns_pos.shape[0] > 0:
+            sns_names = get_sensor_binning(file_name).index
+            pmt_ids   = DB.DataPMT(db_file, run_no).SensorID
+            pmt_name  = sns_names.str.contains('Pmt')
```
Again a comment of what and why is happening would be useful, ie the names are of pmts but no sipms names in sns_names ?

mmkekic replied:
Seems that `get_sensor_binning` gives name 'SiPMSensl' while in the new format is 'SiPM', should we change that?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -275,88 +443,134 @@ def load_mcparticles_df(file_name: str) -> pd.DataFrame:
+                              'configuration').set_index('param_key')
+    sns_names      = load_dst(file_name, 'MC', 'sns_positions').sensor_name
+    if len(sns_names) == 0:
+        return pd.DataFrame(index=['sns_name'], columns=['bin_width'])
```
There can be a file with empty 'sns_position' ? In any case you can read config after you check for the condition

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -275,88 +443,134 @@ def load_mcparticles_df(file_name: str) -> pd.DataFrame:
+    sns_names      = load_dst(file_name, 'MC', 'sns_positions').sensor_name
+    if len(sns_names) == 0:
+        return pd.DataFrame(index=['sns_name'], columns=['bin_width'])
+    sns_binnings   = [nm + '_binning' for nm in sns_names.unique()]
```
Either a comment or in-detail docstring would be useful

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -275,88 +443,134 @@ def load_mcparticles_df(file_name: str) -> pd.DataFrame:
+    return bins[~bins.index.duplicated()]
+
+
+def get_sensor_types(file_name : str) -> pd.DataFrame:
```
is this even used any more?

andLaing replied:
It was a new function that @jmunozv requested. If you think it's better we can remove it from this PR and get him to make a separate one adding the function.

mmkekic replied:
Its ok to be here, but in case this is only valid for new format raise Exception if the old format is passed

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -367,31 +581,43 @@ def load_mcsensor_response_df(file_name : str,
+        if return_raw:
+            return sns_resp
+        return convert_timebin_to_time(sns_resp                          ,
+                                       get_sensor_binning     (file_name),
```
I assume we can not have 'sns_response' without 'sns_position' table? But probably we do need a protection for that?

paolafer replied:
Exactly, 'sns_response' is only filled with the positions of the sensors that have detected some photoelectron, at the moment. It's not a bad idea to add a protection, though, because we may want to save that table with all the sensor positions, in the future.

paolafer replied:
In any case, all the tables are present in the file, always. They may be empty, though.

andLaing replied:
Whether we have all the positions saved or just the ones that are hit doesn't affect this function since we only use the names of the detectors, not the positions etc.

In case of an empty sns_positions table there might be a fail path but I'm not 100% sure what it would be. I'll have a look to see if I can understand it a bit better

mmkekic commented on invisible_cities/cities/components.py:
```diff
@@ -168,20 +174,21 @@ def copy_mc_info(files_in : List[str], h5out: tb.File, event_numbers: List[int])
+                                         if evt in event_numbers_in_file)
+            mcinfo_io.copy_mc_info(f, writer, event_numbers_to_copy,
+                                   db_file, run_number)
+        except tb.exceptions.NoSuchNodeError:
```
This is confusing, when do we expect NoSuchNodeError? Does this mean we skip a file that does not have MC information? This doesnt seem right... Also there is no protection that all `event_numbers` are eventually contained in the output file

mmkekic replied:
I believe that the error can be raised only if the file didnt contain some essential table in which case we do want user to react on that, and not to sneaky catch the exception and continue the work.

mmkekic replied:
On the other hand we never check that all of the `event_numbers` are actually copied, guess we could append `event_numbers_to_copy` in a list and at the end compare that its the same as the `event_numbers`list?

mmkekic reviewed (commented):
All the functions should have more detailed docstrings, especially what type of information is contained in the loaders outputs.
There are bunch of tests still missing and it would be good to add some standardized test files with the same nexus event in both old and new format, for all combinations (full, up to hits, optical photons etc...) , @jmalbos  promised to help with that :)

andLaing committed:
mcinfo_io_test with tests for new MC readers

andLaing committed:
Adapt components.copy_mc_info for the new writer

Adapted test_copy_mc_info_noMC for new structure

Changed components.copy_mc_info to use new mc_writer

Remove old get_mc_info functions in components.cdst_from_files

andLaing committed:
Remove old MC to dictionary readers

andLaing committed:
Add function to differentiate old and new file formats

andLaing committed:
Modify load_mchits_df to be able to read all formats

andLaing committed:
Add function to cast hits df to dictionary

andLaing committed:
Modify load_mcparticles_df to read all file formats

andLaing committed:
Add function to get correct sample binning from file

For both possible formats the function checks
the configuration table and returns a
DataFrame with the sensor names and binning

andLaing committed:
Modify load_mcsensor_response_df to work with both file formats

andLaing committed:
Add function to check sensor types present

andLaing committed:
Add Enum class for MC table types

andLaing committed:
Add function to get the MC tables in file

andLaing committed:
Add function to get list of events in file

andLaing committed:
Update imports for mcinfo_io

andLaing committed:
Add function to load the nexus configuration

andLaing committed:
Add function to load the sensor positions table

andLaing committed:
Add function to load the MC generators table

andLaing committed:
Add function to load the event mapping table

andLaing committed:
Add function to read all MC tables into a dictionary

Uses load functions and returns a dictionary with
keys the MCTableType and values the df

andLaing committed:
Add a writer for the MC info to be copied to the output file

andLaing committed:
Modify mcinfo_io.copy_mc_info to use the new readers/writers

andLaing committed:
Remove obsolete mcinfo_io.mc_info_writer

andLaing committed:
Remove mcinfo_io.mc_info_writer specific tests

andLaing committed:
Remove mc_info_writer test from indexation_test.py

andLaing committed:
Improve test for cast_todict comparison to old method

andLaing committed:
Fix clarifying comments for pandas gymnastics

andLaing committed:
Add exception for old format file in get_sensor_types

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -313,6 +313,47 @@ def load_mcconfiguration(file_name : str) -> pd.DataFrame:
+        if sns_pos.shape[0] > 0:
+            ## Add a column to the DataFrame so all info
+            ## is present like in the new format
+            sns_names = get_sensor_binning(file_name).index
```
This seems like you need a function that just extract sensor_names, for both old and new format from config table.This is also related to `sensor_binning_new` where you need to read 'sensor_positions' to get the sensor names.

andLaing committed:
Improve load_mcconfiguration for all file types

andLaing committed:
Alter get_sensor_binning to use load_mcconfiguration

andLaing committed:
Added warning to load_mcsensor_positions for oldformat

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -321,11 +321,22 @@ def load_mcconfiguration(file_name : str) -> pd.DataFrame:
     config           = load_dst(file_name, 'MC', 'configuration')
     if is_oldformat_file(file_name):
         binning_keys = config.param_key.str.contains('binning')
+        def binning_name(x):
```
how about something like this ?

```
config           = load_dst(file_name, 'MC', 'configuration')
config.param_key = config.param_key.str.replace('.*Pmt.*\_binning.*', 'Pmt_binning')
config.param_key = config.param_key.str.replace('.*SiPM.*\_binning.*', 'SiPM_binning')
config           = config.drop_duplicates('param_key').reset_index(drop=True)
```

mmkekic replied:
Also, we should check if waveforms are present and non-empty, otherwise we should remove any `_binning` parameters from configuration

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -655,72 +655,8 @@ def get_sensor_binning(file_name : str) -> pd.DataFrame:
-    sns_binnings   = [nm + '_binning' for nm in sns_names.unique()]
-    bins           = config.loc[sns_binnings].copy()
+    config         = load_mcconfiguration(file_name).set_index('param_key')
+    bins           = config[config.index.str.contains('binning')].copy()
```
in case config does not contain binning information (ie bins is empty dataframe) I suggest we return empty df and a warning that the information is not present?

mmkekic replied:
No need to `copy`

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -321,29 +321,42 @@ def read_mc_tables(file_in : str                        ,
+        raise tb.exceptions.NoSuchNodeError(f"No MC info in file {file_in}.")
     except IndexError:
-        raise IndexError(f"No event {n} in file {h5in}.")
+        raise IndexError(f"No events {which_events} in file {file_in}.")
```
Remove `IndexError`, that was needed for the previous mc writer to work. Also remove the impossible `NoParticleInfoFile` exception
The `NoSuchNodeError` can happen only in `get_event_numbers_in_file` if none  of the tables are present. Id say remove that too and give string explanation in `get_event_numbers_in_file`as 'at least one of ... tables has to be present in the file'

andLaing committed:
Remove unnecessary index check and raise in read_mc_tables

andLaing committed:
Alter test_copy_mc_info_which_events_out_of_range to new conditions

andLaing committed:
Remove unnecessary try except from mcinfo_io.copy_mc_info

andLaing committed:
Improve error message in _get_list_of_events_new

Raises error if no useful MC tables present

andLaing committed:
Add warning and empty df return to get_sensor_binning

returns empty if no binning info available

andLaing committed:
Add warning to components.copy_mc_info in case of no MC

andLaing committed:
Check warning in test_copy_mc_info_noMC

andLaing committed:
Add old/new format files for 3 types of simulation

Old/new for scintillation only

Old/new for Kr83 fast simulation

Old/new for Kr83 full simulation

andLaing committed:
Add test comparing old and new format for same simulation

andLaing committed:
Add missing column to old format mcparticles

andLaing committed:
Change Pmt binning name for old mcconfiguration

andLaing committed:
Added protection so empty mc sensor positions columns same

andLaing committed:
Test comparing old and new format for same simulaitons

andLaing committed:
Add function to check for presence of MC group

andLaing committed:
Check MC group present and give warning if not for copy_mc_info

andLaing committed:
New test files for mcinfo_test

MC flexible geometry files for mcinfo_io tests

mmkekic commented on invisible_cities/cities/components.py:
```diff
@@ -47,6 +48,7 @@
 from .. io     .pmaps_io          import                load_pmaps
 from .. io     .hits_io           import              hits_from_df
 from .. io     .dst_io            import                  load_dst
+from .. io     .hits_io           import                 load_hits
```
why? XD

mmkekic commented on invisible_cities/cities/components.py:
```diff
@@ -168,20 +175,23 @@ def copy_mc_info(files_in : List[str], h5out: tb.File, event_numbers: List[int])
+        else:
+            warnings.warn(f' File does not contain MC tables.\
+             Use positve run numbers for data', UserWarning)
+            continue
```
I think it would be good at the end of file loop to check that all the events are actually found in the files; basically collect all `event_numbers_to_copy` and make sure that it is the same as `event_numbers` lits

mmkekic commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -90,7 +90,7 @@ def test_esmeralda_filters_events(KrMC_hdst_filename_toy, correction_map_MC_file
         event_info = get_event_info(h5out)
         assert length_of(event_info) == nevt_req
-        MC_num_evs = h5out.root.MC.extents[:]['evt_number']
+        MC_num_evs = load_mchits_df(PATH_OUT).index.levels[0]
```
shouldnt it use `get_event_numbers_in_file`?

mmkekic commented on invisible_cities/cities/penthesilea_test.py:
```diff
@@ -1,9 +1,12 @@
 
 from pytest import mark
 
+from pandas.testing import assert_frame_equal
```
we should switch to the `assert_dataframes_close`

mmkekic commented on invisible_cities/cities/penthesilea_test.py:
```diff
@@ -221,28 +219,41 @@ def test_penthesilea_read_multiple_files(ICDATADIR, output_tmpdir):
+                              hits_in1.index.levels[0]])
+    first_event_out = 4
+    evt_out = hits_out.index.levels[0]
+    print(evt_in, evt_out)
```
forgotten print

mmkekic commented on invisible_cities/cities/penthesilea_test.py:
```diff
@@ -221,28 +219,41 @@ def test_penthesilea_read_multiple_files(ICDATADIR, output_tmpdir):
+    assert all(evt_in[first_event_out:] == evt_out)
+
+    all_hit_in      = pd.concat([hits_in1     ,      hits_in2])
+    pd.testing.assert_frame_equal(all_hit_in.loc[evt_in[first_event_out:]],
```
again - switch to our assert function

mmkekic commented on invisible_cities/cities/penthesilea_test.py:
```diff
@@ -221,28 +219,41 @@ def test_penthesilea_read_multiple_files(ICDATADIR, output_tmpdir):
+                                  hits_out                                )
+    all_particle_in = pd.concat([particles_in1, particles_in2])
+    pd.testing.assert_frame_equal(all_particle_in.loc[evt_in[first_event_out:]],
+                                  particles_out                                )
```
This test is very weird anyway, as I understand we want to check that all the events passed are in output file? @gonzaponte you were the original author of this test?

If I am correct isnt it better to do something like this:
```
input_files = glob(file_in)
events_passed = [14004, 14005, 14006, 14007, 14008, 14009]
all_inp_hits      = pd.concat([load_mchits_df(file_) for file_ in input_files], axis=0)
all_inp_particles = pd.concat([load_mcparticles_df(file_) for file_ in input_files], axis=0)
all_out_hits      = load_mchits_df(file_out)
all_out_particles = load_mcparticles_df(file_out)
assert_frame_equal(all_inp_hits.loc[events_passed, :], all_out_hits)
assert_frame_equal(all_inp_particles.loc[events_passed, :], all_out_particles)
```

ie, instead of hard coding the number of first events from first file that didnt pass filter just hardcode the event numbers that we expect in the output? and later check that the mcinfo is the same?

andLaing committed:
Change esmeralda_test to use get_event_numbers_in_file

andLaing committed:
Replace uses of pandas.testing.assert_frame_equal with IC function

andLaing committed:
Remove unnecessary import from components.py

andLaing committed:
Remove unused imports from mcinfo_io

andLaing committed:
Add MCEventNotFound exception

andLaing committed:
Use MCEventNotFound exception in components.copy_mc_info

If all events not found in loop

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -1,397 +1,766 @@
+                    tbl.file_index.replace(orig_indx, new_indx, inplace=True)
+                except AttributeError:
+                    tbl['file_index'] = first_indx
+            mcwriter_(df=tbl, table_name=key.name)
```
you should specify :
`columns_to_index` - it should be event for most of the relevant tables (hits, particles etc)
`str_col_length` - seems some of the strings are longer than 32 characters, can you identify for which table and then increase the number?

mmkekic commented on invisible_cities/cities/irene_test.py:
```diff
@@ -171,6 +176,7 @@ def test_irene_output_file_structure(config_tmpdir):
         assert "runInfo"      in h5out.root.Run
 
 
+@mark.xfail
```
Instead of skipping can we just set `run_number`to something else and fake 'data' so we dont try to copy MC?

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
+
+    ## Check that we can read the output but it's empty
+    hits = load_mchits_df(file_out)
+    assert hits.shape == (0, 6)
```
maybe hits.empty?

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
+
+    old_tbls = read_mc_tables(file_out_old)
+    new_tbls = read_mc_tables(file_out_new)
+    assert len(old_tbls) == len(new_tbls)
```
why just length and not by element comparison?

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
+                    "nexus_new_kr83m_fast.newformat.sim.h5"),
+                   ("nexus_new_kr83m_full.oldformat.sim.h5",
+                    "nexus_new_kr83m_full.newformat.sim.h5")))
+def test_copy_mc_info_same_results(ICDATADIR   , config_tmpdir,
```
Would be good to split this test in 2, one is just `read_mc_tables`and other copy+read. Also give more descriptive name, ie 'test_copy_mc_info_same_results_old_and_new_format' or so, the idea is to understand what the test does from the test name solely

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
+                                    check_types=False)
+
+
+def test_read_mc_tables(mc_particle_and_hits_nexus_data_new):
```
again change name, this test checks that the tables shape are the same, so more of 'test_read_mc_tables_correct_shape'

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
+        assert shp == tbl_dict[tbl].shape
+
+
+def test_mc_writer(mc_particle_and_hits_nexus_data_new, config_tmpdir):
```
'test_mc_writer_contains_all_tables' (or any more correct version).
Also a good test would be to read, save and read back checking that the read tables are the same. As well as copy_mc_tables comparing that read+write = copy

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
 
-def test_load_mchits(mc_particle_and_hits_nexus_data):
-    efile, _, _, _, _, _, _, X, Y, Z, E, t = mc_particle_and_hits_nexus_data
+def test_mc_writer_oldformat(mc_sensors_nexus_data, config_tmpdir):
```
again rename to be more explicit

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
 
 
-def test_read_mchits_df(mc_particle_and_hits_nexus_data):
+def test_load_mchits_df(mc_particle_and_hits_nexus_data):
```
Either rename the test (`test_load_mchits_df_oldfile`) or do the trick with indirect fixtures.

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -264,62 +280,45 @@ def test_load_mchits_df(mc_particle_and_hits_nexus_data):
-    init_vtx = particle[['initial_x', 'initial_y',
-                         'initial_z', 'initial_t']]
-    assert_allclose(init_vtx.tolist(), vi)
+def test_load_mcparticles_df(mc_particle_and_hits_nexus_data):
```
the same as with `test_load_mchits_df`

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -341,137 +340,99 @@ def test_read_mcparticles_df(mc_particle_and_hits_nexus_data):
-    assert samples[0]   == pmt0_first
-    assert samples[-1]  == pmt0_last
-    assert len(samples) == pmt0_tot_samples
+@fixture(scope = 'module')
```
Maybe you can manually construct the exact binning dataframe you expect for both new and old format (since it is 2-3 rows) so the check is more obvious?

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -206,54 +89,187 @@ def test_copy_mc_info_which_events_out_of_range(ICDATADIR, config_tmpdir):
+    assert MCTableType.sns_response  in saved_tbls
+
+
+@fixture(scope = 'module')
```
if  `mc_particle_and_hits_nexus_data` is only used in this module you can move it here from conftest.py

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -1,397 +1,766 @@
+    """
+    if is_oldformat_file(file_name):
+        raise TypeError('Old format files not valid for get_sensor_types')
+    sns_pos = load_dst(file_name, 'MC', 'sns_positions').copy()
```
no need to `copy`

mmkekic reviewed (commented):
Maybe you can add tests:
- Error when input file contains table outside of McTableType
- Explicit test for `get_event_numbers_in_file` and the Error in case no table is present
- Trivial test for `is_old_format`
- Tests for `load_mcsensor_positions`: for old format that the error is raised; for both that the sensors in sensor_responses are presend in mcsensor_position df
- Tests for `load_mcsensor_response_df` with `return_raw` both `True` and `False` for both formats

andLaing committed:
Change run number for test_empty_events_issue_81

andLaing committed:
Add str_col_length, columns_to_index keyword arguments in mc_writer

andLaing committed:
Updated mcinfo_io_test names

Slight adjustment to comparisons in
test_copy_mc_info_which_events_out_of_range

andLaing committed:
Reset index in load_mcsensor_response_dfold

Makes df index types the same in old and
new format reads

andLaing committed:
Remove unnecessary copies from get_Sensor_binning

andLaing committed:
Add new file with fake MC group for exception tests

andLaing committed:
Add test to check KeyError raised if unrecognised MC table

andLaing committed:
Add tests for get_event_numbers_in_file

andLaing committed:
Add test for is_oldformat_file

andLaing committed:
Add test warning raised without dabtabase in load_mcsensor_positions

Old format file

andLaing committed:
Add test for load_mcsensor_response_df return_raw=True

andLaing committed:
Add database arguments to cop_mc_info call in hypathia

andLaing committed:
Remove MC table comparison from hypathia_test

andLaing committed:
Remove unused imports from mcinfo_io_test

andLaing committed:
Remove unused MCParticle class

andLaing committed:
Remove unused MCParticleInfo class

andLaing committed:
Remove unused get_mc_info_safe

andLaing committed:
Changed get_sensor_binning tests to use DataFrame comparison

andLaing committed:
Move mc_particle_and_hits_nexus_data fixture to mcinfo_io_test

andLaing committed:
Move mc_sensors_nexus_data fixture to mcinfo_io_test

andLaing committed:
Change drop in get_sensor_binning to not use inplace

Avoids warning/error from pandas

mmkekic reviewed (commented):
in all city tests remove comments and replace the tables with new MC tables we expect for 2 types of test:
- check if table present
- exact comparison. for this one you have to replace the exact_comparison file

After this and some squashing the PR is ready to be approved (finally)

andLaing committed:
Add new test files for city exact results

andLaing committed:
Use NEWMC files in cities exact result tests. All tables compared

andLaing committed:
Protec get_sensor_binning from merged files

mmkekic reviewed (approved):
This PR modifies MC io module to adapt for the new nexus output format, ensuring that the main readers/writer can be applied to the old format as well. The new functions use extensively dataframes that are faster to manipulate with than the previous dictionary based readers/writer. All the functionalities are well tested, and new MC test dataset is introduced to guarantee the equality in information read from the old and new format. 
Some functionalities are repeated due to different old/new table structures, and are planned to be removed in the future once there is no need for the 'old' files readers.
The road to user-transparent compatibility was bumpy but it sparks joy to finally approve this PR! Well done!
